---
title: "Coral Heat Stress"
author: "Grace Beery, Hayden Dickerson, Susritha Kopparapu"
date: "3/27/2021"
output: 
  html_document:
    fig_width: 6 
    fig_height: 4 
---
*The following script is modified from Sarah Davies and James Fifer*

## Introduction
Due to climate change, water temperatures and atmospheric carbon dioxide levels are projected to increase significantly in future years. Calcifying organisms, such as corals, are expected to be the most drastically impacted by heat and acidification stress due to their fragile structures. This study looks at a resilient reef-building coral, *Siderastrea siderea*, to determine how heat stress and acidification affect its molecular pathways. For our reanalysis, we are looking specifically at heat stress versus control. Our subsample contained 1 control sample (28 degrees C) from the nearshore region, 1 control sample from the fore-reef region, 1 heat sample (32 degrees C) from the nearshore region, and 1 heat sample from the fore -reef region. Each sample was kept at a standard atmospheric CO2 level of between 472-477 Î¼atm. All data came from the [Thermal and pCO2 stress elicit divergent transcriptomic responses in a resilient coral](https://www.frontiersin.org/articles/10.3389/fmars.2016.00112/full) study (Davies et al. 2016). 

## Version Control

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### R Version
R version 4.0.2 was used for this analysis. 

#### Packages
The following packages were used to clean, analyze, and visualize data. Packages were installed with [Bioconductor](https://bioconductor.org/biocLite.R) version 3.12.  
```{r, message=FALSE}
library("affycoretools")            #Version 1.60.1
library("arrayQualityMetrics")      #Version 3.44.0
library("genefilter")               #Version 1.70.0
library("DESeq2")                   #Version 1.28.1
library("ggplot2")                  #Version 3.3.3
library("dplyr")                    #Version 1.0.1
library("pheatmap")                 #Version 1.0.12
library("vegan")                    #Version 2.5.7
library("ggrepel")                  #Version 0.8.2
library("tidyverse")                #Version 1.3.0
library("ape")                      #Version 5.4.1
library("gplots")                   #Version 3.0.4
```

```{r, eval=FALSE, echo=FALSE}
packageVersion("affycoretools")
packageVersion("arrayQualityMetrics")
packageVersion("genefilter")
packageVersion("DESeq2")
packageVersion("ggplot2")
packageVersion("dplyr")
packageVersion("pheatmap")
packageVersion("vegan")
packageVersion("ggrepel")
packageVersion("tidyverse")
```

## Setup
```{r, echo=FALSE}
setwd("/usr4/bi594/skoppara/Assignment2/Assign2_gr2")
```

We first read in the isogroup gene annotations file and the raw counts from the study (Davies et al. 2016). There are a total of 16931 isogroups in this data set. 
```{r}
gg=read.table(file="davies_Ssid_iso2gene.tab",sep="\t", quote="")     #must specify quote="", or EOF error occurs
countData <- read.csv("2.Assign2_GE_Coral_OA.csv")
head(countData)
length(countData[,1])
```

The first column of the `countData` is the names of the isogroups, so the first line of code is used to turn the first column into the row names. Then the column names of `countData` are set to be the values of the character vector. In this data frame there are two control samples and two hot samples. 
```{r}
countData <- data.frame(countData, row.names = 1)                     #NOTE: this line should only run once
names(countData)=c( "Control_1", "Hot_1", "Control_2", "Hot_2")
head(countData)
```

The dataframe is now properly set to check for outliers. 

## Outlier Analysis 
A vector is set up so that it can be converted to a data frame with the names of the different treatments. These correspond to the column names.
```{r}
treat=c( "Control_1", "Hot_1", "Control_2", "Hot_2")
colData=data.frame(treat)
colData
```

The countData is set to the raw counts, the colData is set to the names of each of the columns, and the design is set to the character vector treat, that holds the names of the treatments. The DESeqDataSetFromMatrix step is run.
```{r, warning=FALSE}
dds=DESeqDataSetFromMatrix(countData=countData, colData=colData, design=~treat)
```

The `vst` function is the variance-stabilizing transformation which models the mean-variance relationship of replicates. The assay of `r1` generates a table with the mean-variance of each of the isogroup treatments. The `as.data.frame(colData(rl)))` generates a table with the columns and a sizeFactor value that is converted into the formal class AnnotatedDataFrame. Both values are passed are used to create an `ExpressionSet` object which is used to generate the array's quality metrics using the `treat` variable as the intgroup. 
```{r, eval=FALSE}
rl=vst(dds)
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
arrayQualityMetrics(e,outdir=v,intgroup=c("treat"),force=T)
```

### Visualization
The array's quality metrics showed that there were no outliers by the three detection methods: by distances between arrays, by boxplots, and by MA plots. Below is the heatmap of sample distance between arrays. 


![**Figure 1.** Heatmap of the distances between arrays generated with array quality metrics. The color scale is chosen to cover the range of distances encountered in the dataset. Patterns in this plot can indicate clustering of the arrays either because of intended biological or unintended experimental factors (batch effects). No outliers for any of the four arrays were detected.](/usr4/bi594/skoppara/Assignment2/Assign2_gr2/images/hm.png)

## Raw Counts Analysis

Next, a data frame made of the treatments and the raw counts is used to generate a bar plot of total counts. 
```{r}
summedCounts=as.data.frame(colSums(countData))
totalCounts = as.data.frame(treat)
totalCounts$raw = summedCounts[,1]
totalCounts$treat <- factor(totalCounts$treat, levels = totalCounts$treat)

```

```{r, echo=FALSE, fig.cap = "**Figure 2.** Bar plot of the raw totals of the number of isogroups for each of the treatment temperatures in the *Siderastrea siderea*." }
p <- ggplot(totalCounts, aes(x = treat, y = raw))
p + geom_bar(aes(fill = treat), stat="identity") + xlab("Temperature") + ylab("Raw Counts")
```


Below are the totals, minumum, and maximum of the raw counts. 
```{r, include=FALSE}
totalCounts=colSums(countData)

```

```{r}
totalCounts
min(totalCounts)
max(totalCounts)
```

## DESeq Analysis 

A vector of the temperatures of each sample, and vector of the genotype of each sample is made and the `DESeqDataSetFromMatrix` step is repeated with the treatment as the temperature alone. Then that large DESeqDataSet is handed to `DESeq`. 
```{r, warning=FALSE, message=FALSE}

temperature=c( "Control", "Hot", "Control", "Hot")
genotype=c( "a", "a", "b", "b")
colData=data.frame(temperature, genotype)
dds<-DESeqDataSetFromMatrix(countData=countData, colData=colData, design=~temperature) 
dds<-DESeq(dds)

```


`results` extracts the results table from the DESeq analysis and the DESeqDataSet is used to generate the dispersion plot. 
```{r}
head(dds)
res<- results(dds)
```

```{r, echo=FALSE, fig.cap="**Figure 3.** Scatter plot of dispersion estimates per gene (on the y-axis) versus the mean of normalized counts (on the x-axis) of isogroups identified in the treatment samples of *Siderastrea siderea*. Gene-wise dispersion estimates are in black, fitted estimates are in red, and final estimates are in blue as indicated by the legend. "}
plotDispEsts(dds, main="Dispersion plot *Siderastrea siderea*")

```

## Hot vs. Control Pairwise Comparison

The temperature column is represented as a factor with two levels being either hot or control. These are stored in a new column called heat. Then the results function extracts the results table of DESeqDataSet with the contrast set as temperature, between hot and control.
```{r}
colData$heat<-factor(colData$temperature, levels=c("Hot", "Control"))
resheat <- results(dds, contrast=c("temperature","Hot","Control"))

```

The results table reports a p-adjusted value for every isogroup, and below is the breakdown of how many of these isogroups had p-adjusted values below 0.01, to be considered significant. The summary of the results with a contrast by temperature is also displayed below. Notice that while only 2.3% of genes were upregulated, 11% were downregulated. 
```{r}
table(resheat$padj<0.01)
summary(resheat)
```

The number of genes that have a p-adjusted value less than 0.05 and that do not have NA values for the p-adjusted values, are counted and reported below. 
```{r}
nrow(resheat[resheat$padj<0.05 & !is.na(resheat$padj),])
```

```{r, echo=FALSE, fig.cap="**Figure 4.** Hot vs. Control temperatures scatter plot of log fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis) of the isogroups identified in the treatments samples of *Siderastrea siderea*. The non signicant data points are in gray, and significant data points are represented in blue."}
plotMA(resheat, main="Hot vs Control", ylim=c(-2,2))
```

```{r, include=FALSE}
results <- as.data.frame(resheat)
```


The number of upregulated and downregulated genes that have a p-adjusted value less than 0.1 and that do not have NA values for the p-adjusted values, are counted and reported below.
```{r}
nrow(resheat[resheat$padj<0.1 & resheat$log2FoldChange > 0 & !is.na(resheat$padj),])
nrow(resheat[resheat$padj<0.1 & resheat$log2FoldChange < 0 & !is.na(resheat$padj),])
```

```{r, include = FALSE}
write.table(resheat, file="resheat.txt", quote=F, sep="\t")
cd <- read.table("resheat.txt")
head(cd)
```


## GO table for MWU

Use the results from the DESeq analysis with the contrast as the temperature to generate the GO table for the specific isogroups in the samples. This generates the table with the signed p values that can be used for the GO enrichment analysis later on. 
```{r}
go_input_heat = cd %>%
  tibble::rownames_to_column(var = "iso") %>%
  mutate(mutated_p = -log(pvalue)) %>%
  mutate(mutated_p_updown = ifelse(log2FoldChange < 0, mutated_p*-1, mutated_p*1)) %>%
  na.omit() %>%
  select(iso, mutated_p_updown)
colnames(go_input_heat) <- c("gene", "pval")
head(go_input_heat)
write.csv(go_input_heat, file="heat_GO.csv", quote=F, row.names=FALSE)
```

## P values and Rlogdata
The p-values and p-adjusted values are bound together and the number of rows representing the isogroups were preserevd when checked against the p-values.
```{r}
valheat=cbind(resheat$pvalue, resheat$padj)
colnames(valheat)=c("pval.heat", "padj.heat")
length(valheat[,1])
```

Below is a report of the number of complete cases that were not missing any values. 
```{r}
table(complete.cases(valheat))
```

R log transformation is used on the data as a data normalization method. This r log tranformation will be important for making heatmaps. The last line sanity checks length to ensure we are not losing any data.
```{r}
rlog=rlogTransformation(dds, blind=TRUE) 
rld=assay(rlog)
length(rld[,1])
```

Then, we bind together the rlog data and p values. This allows us to be able to sort values for significance when plotting heatmaps.
```{r}
rldpvals=cbind(rld,valheat)
dim(rldpvals)
```

The table of the combined rlog data and p values should remain the same number of FALSE / TRUE as our p values above. 
```{r}
table(complete.cases(rldpvals))
```

## Heat map of sample distances
We plot a sample distance heatmap to see how general expression compares from sample to sample. The purpose of this map is to look at how overall expression from each sample is similar to each other, and can be used as a sanity check to ensure our data is being displayed in the way that we expect it to be. As we see the controls are more similar to the controls and the hot treatments are more similar to themselves, we can conclude that our data looks how we would expect and can proceed.
```{r}
sampleDists <- as.matrix(dist(t(rld)))
```

```{r, echo = FALSE, fig.cap="**Figure 5.** A sample distance matrix heatmap to see the general expression comparison from sample to sample." }
heatmap.2(as.matrix(sampleDists), key=F, trace="none",
          col=colorpanel(100, "black", "white"),
          margin=c(10, 10))
```

We create a table containing our previously normalized data and p values, but exclude the last two columns to ensure p values are excluded from our heatmap plots. While p values are informative for helping pick out which genes we want to visualize, we don't actually want to visualize the p value data.
```{r}
rldpvals <- read.csv(file="resheat_RLDandPVALS.csv", row.names=1)
rld=rldpvals[,1:4]
sampleDists <- dist(t(rld))
sampleDistMatrix <- as.matrix( sampleDists )
treat=c( "Hot", "Control")
```

Using the table above, a heatmap is created. This is essentially the same heatmap as our Sample Distance Matrix above, just in color.
```{r,  echo = FALSE, warning=FALSE, message=FALSE, fig.cap="**Figure 6.** A sample distance matrix heatmap to see the general expression comparison from sample to sample in color."}
heat.colors = colorRampPalette(rev(c("blue","yellow","red")),bias=0.3)(100)
pheatmap(sampleDistMatrix,color = heat.colors,cex=0.9,border_color=NA,cluster_rows=T,cluster_cols=T)

```


## PCA
Principal component analysis of the r-log transformed data.
```{r}
rld_t=t(rld)
pca <- prcomp(rld_t,center = TRUE)
li <- pca$sdev^2 / sum(pca$sdev^2)
pc1v <- round(li[1] * 100, 1)
pc2v <- round(li[2] * 100, 1)
pca_s <- as.data.frame(pca$x)
pca_s <- pca_s[,c(1,2)]
pca_s$Samples = row.names(pca_s)
pca_s$treat=colData$temperature
```

Here, we are essentially looking at the distance between the two points, to visualize overall variance. Variance on the Y axis is explained by the different study site locations, ie. fore reef versus nearshore. 
```{r,  echo = FALSE, fig.cap="**Figure 7.** Principal component analysis of all samples in the subset of data where the PC1 along the x-axis is the only meaningful axis of variance. The legend shows the difference between control and hot treatment samples. See below adonis data that shows this variance is significant."}
cbPalette <- c("darkgoldenrod2",  "darkolivegreen3") #, "dodgerblue3", "red")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
```

We then use `adonis` to determine if gene expression between treatments is significant or not. We see that the treatment plotted has an R2 value of .91461, which can be considered significant.
```{r, warning=FALSE, message=FALSE}
adonis(pca$x ~ treat, data = pca_s, method='eu', na.rm = TRUE)
```

## Heat map for genes 

This heatmap will display the differences in gene expression for each treatment based on their site of origin. Here, we are looking for similarities in gene expression.

```{r}
rldpvals <- read.csv(file="resheat_RLDandPVALS.csv", row.names=1)
rld_site= rldpvals[,1:4]
gg=read.table("davies_Ssid_iso2gene.tab",sep="\t", row.names=1, quote ='')
```

```{r}
nrow(rldpvals[rldpvals$padj.heat<0.01& !is.na(rldpvals$padj.heat),])
```

topnum = 100 represents the number of differentially expressed genes. 
```{r}
topnum= 100
top100=head(rldpvals[order(rldpvals$padj.heat), ],topnum)
length(top100[,1])
summary(top100)
```

We use a p value of 0.1 as our FDR cutoff. 
```{r}
p.val=0.1 
conds=top100[top100$padj.heat<=p.val & !is.na(top100$padj.heat),]
length(conds[,1])

exp=conds[,1:4] 
means=apply(exp,1,mean) 
explc=exp-means 
```

Assign colors and create a heatmap. Here, we are looking for similarities in gene expression between hot 1 and hot 2, and between control 1 and control 2 to ensure that difference in study side (fore reef vs nearshore) does not significantly influence gene expression.
```{r,  echo = FALSE, fig.cap="**Figure 8.** A heatmap of similarities in gene expression between hot1 and ho2 and control1 and control 2."}
ccol=colorRampPalette(rev(c("red","chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)

pheatmap(explc,cluster_cols=T,scale="row",color=ccol, show_rownames = F)
```

## Heat map for genes in common 

Our next heatmap will display, among the genes in commmon for all treatments, upregulation and downregulation based on treament. The p-value for FDR cutoff was chosen to be 0.1
```{r}
rldpvals <- read.csv(file="resheat_RLDandPVALS.csv", row.names=1)
p.val=0.1 # FDR cutoff
conds=rldpvals[rldpvals$padj.heat<=p.val & !is.na(rldpvals$padj.heat),]
rld_data= conds[,c(1:4)]
gg=read.table("davies_Ssid_iso2gene.tab",sep="\t", row.names=1, quote='')
means=apply(rld_data,1,mean) # means of rows
explc=rld_data-means
```

Assign colors and create heatmap. Here, we can see that the genes that are upregulated in the control are now downregulated in the heat treatment, and vice versa.Out of the genes that our control and heat treatment have in common, 379 are upregulated in the heat treatment, and 1839 are downreguated in the heat treatment.
```{r,  echo = FALSE, fig.cap="**Figure 9.** A heatmap of the genes in common that shows the genes that were upregulated and downregulated."}
ccol=colorRampPalette(rev(c("red","chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)

pheatmap(explc,cluster_cols=T,scale="row",color=col0, show_rownames = F)
```



## GO Enrichment Analysis 

The *Siderastrea sidera* transcriptome isogroup GO annotations were retrieved from the study (Davies et al. 2016) The GO database was downloaded from [Gene Ontology](http://geneontology.org/docs/download-ontology/) and here the go_basic.obo file is being used, but was renamed go.obo. This analysis was done for the biological processes division. 

Furthermore, all `gomwustats` settings were made to default, except the `smallest` parameter that was raised to 10 to set more stringent boundaries. 
```{r, eval=FALSE}
input="heat_GO.csv"
goAnnotations="davies_Ssid_iso2go.tab" 
goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDivision="BP" 
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="C:/Strawberry/perl/bin/perl.exe", 
           largest=0.1, 
           smallest=10,   
           clusterCutHeight=0.25
)
```

As there was a strong signal, the levels were chosen to be very stringent as described in the caption.  
```{r, eval=FALSE, echo=FALSE}
results=gomwuPlot(input,goAnnotations,goDivision,
	absValue=1,
	level1=0.01, 
	level2=0.001,
	level3=0.0001, 
	txtsize=1.2,    
	treeHeight=1, 
  colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") 
)
```
![**Figure 10.** GO MWU plot with level 1 at 0.01, level 2 at 0.001, and level 3 at 0.0001. The biological processes (BP) sub-ontology was used for this analysis. Blue represents upregulated GO terms and red represents downregulated GO terms](/usr4/bi594/skoppara/Assignment2/Assign2_gr2/images/goplot.png)

-----------------------

## Conclusion 
Overall, we found that heat stress significantly affects molecular pathways in *Siderastrea siderea*. Our findings were mostly consistent with the findings of the original study, but more data is necessary to confirm our findings. We found that the heat treated corals have significant differences in gene expression-- the original study found that 72.3% of DEGs were underrepresented in the heat stress treatment (Davies et al. 2016), whereas our reanalysis found that close to 83% of DEGS were underrepresented in the heat stressed corals. This discrepancy could be due to the fact that analyses from the original to the reanalysis were performed in different versions of R, and that our reanalysis contained significantly fewer variables than the original, which could have been confounding the results. We also saw that in our GO enrichment analysis for Biological Processes, catabolic pathways are significantly upregulated under heat stress, which is consistent with the findings of the original study. Decreases in metabolic activity are known to be an adaptive strategy for managing short-term stress, however, long term upregulation of catabolic processes can lead to protein breakdown, so its possible that the corals under 32 degree Celsius heat stress are reaching the threshold of heat stress that they are unable to recover from. Overall, our findings largely support the finidngs of the original study, and emphasize how detrimental rising sea temperatures can be for reef-building corals, even resilient corals like *Siderastrea siderea*.

## Reference
Davies SW, Marchetti A, Ries, JB and KD Castillo (2016) Thermal and pCO2 stress elicit divergent transcriptomic responses in a resilient coral. Frontiers in Marine Science. FMARS-03-00112.